vsintegration/src/FSharp.Editor/FsiCommandService.fs

```diff
@@ -0,0 +1,92 @@
+﻿namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.OLE.Interop
+open System.ComponentModel.Composition
+open Microsoft.VisualStudio
+open Microsoft.VisualStudio.Editor
+open Microsoft.VisualStudio.TextManager.Interop
+open Microsoft.VisualStudio.Utilities
+open Microsoft.VisualStudio.Shell
+open Microsoft.VisualStudio.Shell.Interop
+open Microsoft.VisualStudio.FSharp.Interactive
+open EnvDTE
+
+type IMenuCommand =
+    inherit IOleCommandTarget
+    abstract IsAdded: bool with get, set
+    abstract NextTarget: IOleCommandTarget with get, set
+
+type FsiCommandFilter(serviceProvider: System.IServiceProvider) =
+    let projectSystemPackage =
+      lazy(
+        let shell = serviceProvider.GetService(typeof<SVsShell>) :?> IVsShell
+        let packageToBeLoadedGuid = ref (Guid "{91a04a73-4f2c-4e7c-ad38-c1a68e7da05c}")  // FSharp ProjectSystem guid        
+        match shell.LoadPackage(packageToBeLoadedGuid) with
+        | VSConstants.S_OK, pkg ->
+            pkg :?> Package
+        | _ -> null)
+
+    interface IMenuCommand with
+        member val IsAdded = false with get, set
+        member val NextTarget = null with get, set
+
+        member x.Exec (pguidCmdGroup, nCmdId, nCmdexecopt, pvaIn, pvaOut) =
+            if pguidCmdGroup = VSConstants.VsStd11 then
+                if nCmdId = uint32 VSConstants.VSStd11CmdID.ExecuteSelectionInInteractive then
+                    Hooks.OnMLSend projectSystemPackage.Value FsiEditorSendAction.ExecuteSelection null null
+                elif nCmdId = uint32 VSConstants.VSStd11CmdID.ExecuteLineInInteractive then
+                    Hooks.OnMLSend projectSystemPackage.Value FsiEditorSendAction.ExecuteLine null null
+                VSConstants.S_OK
+            elif pguidCmdGroup = Guids.guidInteractive then
+                if nCmdId = uint32 Guids.cmdIDDebugSelection then
+                    Hooks.OnMLSend projectSystemPackage.Value FsiEditorSendAction.DebugSelection null null
+                VSConstants.S_OK
+            else
+                let x = x :> IMenuCommand
+                x.NextTarget.Exec(&pguidCmdGroup, nCmdId, nCmdexecopt, pvaIn, pvaOut)
+
+        member x.QueryStatus (pguidCmdGroup, cCmds, prgCmds, pCmdText) =
+            if pguidCmdGroup = VSConstants.VsStd11 then
+                if prgCmds.[0].cmdID = uint32 VSConstants.VSStd11CmdID.ExecuteSelectionInInteractive then
```

Comment:
I only encountered one-element `prgCmds` but one can't guarantee that. I made a relevant change in cd94a4e.